quarkchain/evm/state.py

```diff
@@ -92,49 +92,120 @@ class TokenBalancePair(rlp.Serializable):
 
 
 class TokenBalances:
-    """interface for token balances
-    TODODLL: store token balances in trie when TOKEN_TRIE_THRESHOLD is crossed
-    """
+    """Token balances including non-genesis native tokens."""
 
     def __init__(self, data: bytes, db):
-        self.token_trie = SecureTrie(Trie(db))
-        self.balances = {}
-        self.enum = b"\x00"
+        self.token_trie = None
+        self.token_root = None
+        # We don't want outside world to know this
+        self._balances = {}
+        self._db = db
+        self._committed = False
+
         if len(data) != 0:
-            self.enum = data[:1]
-            if self.enum == b"\x00":
+            enum = data[:1]
+            if enum == b"\x00":
                 for p in rlp.decode(data[1:], CountableList(TokenBalancePair)):
-                    self.balances[p.token_id] = p.balance
-            elif self.enum == b"\x01":
-                raise Exception("Token balance trie is not yet implemented")
+                    self._balances[p.token_id] = p.balance
+            elif enum == b"\x01":
+                self.token_root = data[1:]
+                self.token_trie = SecureTrie(Trie(db, data[1:]))
             else:
-                raise Exception("Unknown enum byte in token_balances")
+                raise ValueError("Unknown enum byte in token_balances")
+
+    def commit(self):
+        self._committed = True
+        # No-op if not using trie and token size is small
+        if not self.token_trie and self._non_zero_entries <= TOKEN_TRIE_THRESHOLD:
+            return
+        # Use trie if size exceeds threshold
+        if not self.token_trie:
+            self.token_trie = SecureTrie(Trie(self._db))
+
+        keys_to_del = []
+        for token_id, bal in self._balances.items():
+            k = utils.encode_int32(token_id)
+            if bal:
+                self.token_trie.update(k, rlp.encode(bal))
+            else:
+                self.token_trie.delete(k)
+                keys_to_del.append(token_id)
+        for token_id in keys_to_del:
+            del self._balances[token_id]
+
+        self.token_root = self.token_trie.root_hash
 
     def serialize(self):
-        if len(self.balances) == 0:
+        if not self._committed:
```

Comment:
Following the previous comment, I think we could check the commit status by checking
- token_trie is None or
- token_trie is not None and _balances = {}

Reply:
updated. should check

1. token trie is None **and has less than 16 entries in balance cache**
2. token trie is not None and balance cache is {}